lib/rules/files.js

```diff
@@ -35,7 +35,10 @@ const process = (props) => {
 
         return true;
       })
-      .sort((a, b) => (a.value > b.value ? 1 : a.value < b.value ? -1 : 0));
+      // sort alphabetically
+      .sort((a, b) => (a.value > b.value ? 1 : a.value < b.value ? -1 : 0))
+      // move the negations (!) to the bottom
+      .sort((a, b) => b.value.startsWith('!') ? -1 : 0);
```

Comment:
I was just about to approve this when I had the notion to run a test:

```js
['!x', '!a', '!g'].sort((a, b) => b.startsWith('!') ? -1 : 0);
// [ '!g', '!a', '!x' ]
```

That's probably not what we're looking for. Since the order of negations is important, we'll want to preserve that order. What we'll probably want to do here instead is `filter` any negations first, into a variable, then filter out the negations from `files` and sort that, and then concat the negations in their original order.

If negation order _doesn't_ matter then we should alphabetize them using a similar method as mentioned above, but comparing the negation strings without the leading `!`.

I'll work to get some tests setup in the repo to account for things like this now that they're getting more complex, which should aid adding features like this moving forward.

Reply:
This is one area where ES6 doesn't have a good builtin mechanism. Some pseudo code:

```js
const negations = [];
const files = elements.filter((element) => {
  if (element.value.startsWith('!')) return !negations.push(element);
  return true;
}).sort(.....);

negations.sort(....)
files.concat(negations);
```

Reply:
What about to just ignore the `!` to make the sort?
I think it could work, and it keeps the negation close to the origin.

Update to sort to:

```javascript
.sort((a, b) => {
  const aValue = a.value.startsWith('!') ? a.value.substring(1) : a.value;
  const bValue = b.value.startsWith('!') ? b.value.substring(1) : b.value;

  return aValue.localeCompare(bValue);
});
```

I have also uses the [`localeCompare()` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare).

Reply:
From this:

```json
  "files": [
    "/lit-html.js",
    "/lit-html.js.map",
    "/lit-html.d.ts",
    "/lit-html.d.ts.map",
    "/directives/",
    "/parts.js",
    "/parts.js.map",
    "/parts.d.ts",
    "/parts.d.ts.map",
    "/src/",
    "!/src/test/",
    "/development/",
    "!/development/test/"
  ],
```

We sort like:

```json
  "files": [
    "/development/",
    "!/development/test/",
    "/directives/",
    "/lit-html.d.ts",
    "/lit-html.d.ts.map",
    "/lit-html.js",
    "/lit-html.js.map",
    "/parts.d.ts",
    "/parts.d.ts.map",
    "/parts.js",
    "/parts.js.map",
    "/src/",
    "!/src/test/"
  ],
```